### PR TITLE
Feature/rde 431 loop upgrade

### PR DESCRIPTION
**Pull request by BillyMurrayOlive**
First pass at the services. Want to get approval on these before moving forward.

- New services for each sensor
- Split each service into a separate file

### CODE REVIEW

**Review by corylanou — APPROVED** [review]

**Review by iangallOlive — APPROVED** [review]

File: proto/process.proto [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    // stream updates to processes as they happen
+    rpc ProcessDiffStream (google.protobuf.Empty) returns (stream ProcessDiffStreamResponse);
+
+    // get a list of all processes
+    rpc ProcessList (google.protobuf.Empty) returns (ProcessListResponse);
+}
+
+message ProcessInfo {
+    int32 pid = 1;
+    string command = 2;
```
Comment (richseviora) [inline proto/process.proto:17]: We could also include the process arguments; not sure if that'll be needed or not

File: proto/window.proto [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    int32 width = 6;
+    int32 height = 7;
+}
+
+enum WindowAction {
+    WINDOW_ACTION_UNKNOWN = 0; 
+    WINDOW_ACTION_FOCUSED = 1;
+    WINDOW_ACTION_UNFOCUSED = 2;
+    WINDOW_ACTION_OPENED = 3;
+    WINDOW_ACTION_CLOSED = 4;
```
Comment (richseviora) [inline proto/window.proto:36]: A window can both focus and open in the same action; it doesn't need to be captured as a separate step here but we should be explicit about how a new window that takes focus is represented (one focus event, or one focus + one open event)

**Review by JonBernardOlive — APPROVED** [review]

**Review by richseviora — APPROVED** [review]
Additional thoughts added :)

File: proto/chrome.proto [hunk 3]
```diff
@@ -0,0 +1,36 @@
+syntax = "proto3";
+package proto;
+option go_package = ".;proto";
+
+import "google/protobuf/empty.proto";
```
[the remark this answers is not in the document]
Comment (BillyMurrayOlive) [reply proto/chrome.proto]: Makes sense to me. I'll rename this from `chrome` to `browser`.

File: proto/chrome.proto [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+message ChromeActiveURLResponse {
+    string url = 1;
+}
+
+message ChromeActiveURLStreamResponse {
+    string url = 1;
+}
+
+message ChromeSelectedTextResponse {
+    string text = 1;
```
[the remark this answers is not in the document]
Comment (BillyMurrayOlive) [reply proto/chrome.proto]: Makes sense. Changes made.
Does it make sense to also add `tabTitle` to the BrowserActiveURLResponse and BrowserActiveURLStreamResponse?

File: proto/cursor.proto [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+service Cursor {
+    // get the current position of the cursor
+    rpc CursorPosition (google.protobuf.Empty) returns (CursorPositionResponse);
+
+    // get the position of the cursor every time it changes
+    rpc CursorPositionStream (google.protobuf.Empty) returns (stream CursorPositionStreamResponse);
+}
+
+message CursorPositionResponse {
+    int32 x = 1;
```
[the remark this answers is not in the document]
Comment (BillyMurrayOlive) [reply proto/cursor.proto:16]: Sure. Not exactly sure what type to make the screen. Making it `uint32 ` for now.

File: proto/hover.proto [hunk 6]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
+package proto;
+option go_package = ".;proto";
+
+import "google/protobuf/empty.proto";
+
+service Hover {
+    // read the text under the cursor
+    rpc HoverRead (google.protobuf.Empty) returns (HoverReadResponse);
```
[the remark this answers is not in the document]
Comment (BillyMurrayOlive) [reply proto/hover.proto:9]: Yes, thanks for the suggestion. Please let me know if there are other opportunities you see for configurable options.

File: proto/keyboard.proto [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+service Keyboard {
+    // stream each scancode as it is pressed
+    rpc KeyboardScancodeStream (google.protobuf.Empty) returns (stream KeyboardScancodeStreamResponse);
+
+    // stream chunks of text when the user finishes typing them
+    rpc KeyboardTextChunkStream (google.protobuf.Empty) returns (stream KeyboardTextChunkStreamResponse);
+
+    // stream text as it is typed
+    rpc KeyboardTextStream (google.protobuf.Empty) returns (stream KeyboardTextStreamResponse);
```
[the remark this answers is not in the document]
Comment (BillyMurrayOlive) [reply proto/keyboard.proto:15]: Thanks for pointing that out. I completely neglected to consider meta keys. Made the following changes:

- KeyboardScancodeStreamResponse added a bool field `pressed`. true = key down, false = key up.
- Added `modifiers` field to KeyboardTextStreamResponse
- New rpc to setup hotkeys

File: proto/process.proto (hunk 1 above)
Comment (BillyMurrayOlive) [reply proto/process.proto:17]: I think it makes sense to add.

File: proto/window.proto (hunk 2 above)
Comment (BillyMurrayOlive) [reply proto/window.proto:36]: I added a comment about this. Let me know if you think it needs more.

**Pull request merged**